pkg/controllers/externalsecret/externalsecret_controller_test.go

```diff
@@ -1474,6 +1474,347 @@ var _ = Describe("ExternalSecret controller", func() {
+		ExternalSecretStore            = "test-cluster-store"
+		ExternalSecretTargetSecretName = "test-secret"
+		NamespaceLabelKey              = "css-test-label-key"
+		NamespaceLabelValue            = "css-test-label-value"
+	)
+
+	var ExternalSecretNamespace string
+
+	// if we are in debug and need to increase the timeout for testing, we can do so by using an env var
+	if customTimeout := os.Getenv("TEST_CUSTOM_TIMEOUT_SEC"); customTimeout != "" {
+		if t, err := strconv.Atoi(customTimeout); err == nil {
+			timeout = time.Second * time.Duration(t)
+		}
+	}
+
+	BeforeEach(func() {
+		var err error
+		ExternalSecretNamespace, err = ctest.CreateNamespaceWithLabels("test-css-ns", k8sClient, map[string]string{NamespaceLabelKey: NamespaceLabelValue})
+		Expect(err).ToNot(HaveOccurred())
+		metric.Reset()
+		syncCallsTotal.Reset()
+		syncCallsError.Reset()
+		externalSecretCondition.Reset()
+		externalSecretReconcileDuration.Reset()
+		fakeProvider.Reset()
+	})
+
+	AfterEach(func() {
+		Expect(k8sClient.Delete(context.Background(), &esv1beta1.ClusterSecretStore{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ExternalSecretStore,
+			},
+		})).To(Succeed())
+		Expect(k8sClient.Delete(context.Background(), &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ExternalSecretNamespace,
+			},
+		})).To(Succeed())
+	})
+
+	const secretVal = "some-value"
+	const targetProp = "targetProperty"
+	const remoteKey = "barz"
+	const remoteProperty = "bang"
+
+	makeDefaultTestcase := func() *testCase {
+		return &testCase{
+			// default condition: es should be ready
+			checkCondition: func(es *esv1beta1.ExternalSecret) bool {
+				cond := GetExternalSecretCondition(es.Status, esv1beta1.ExternalSecretReady)
+				if cond == nil || cond.Status != v1.ConditionTrue {
+					return false
+				}
+				return true
+			},
+			checkExternalSecret: func(es *esv1beta1.ExternalSecret) {
+				// noop by default
+			},
+			secretStore: &esv1beta1.ClusterSecretStore{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ExternalSecretStore,
+				},
+				Spec: esv1beta1.SecretStoreSpec{
+					Provider: &esv1beta1.SecretStoreProvider{
+						AWS: &esv1beta1.AWSProvider{
+							Service: esv1beta1.AWSServiceSecretsManager,
+						},
+					},
+				},
+			},
+			externalSecret: &esv1beta1.ExternalSecret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      ExternalSecretName,
+					Namespace: ExternalSecretNamespace,
+				},
+				Spec: esv1beta1.ExternalSecretSpec{
+					SecretStoreRef: esv1beta1.SecretStoreRef{
+						Name: ExternalSecretStore,
+						Kind: esv1beta1.ClusterSecretStoreKind,
+					},
+					Target: esv1beta1.ExternalSecretTarget{
+						Name: ExternalSecretTargetSecretName,
+					},
+					Data: []esv1beta1.ExternalSecretData{
+						{
+							SecretKey: targetProp,
+							RemoteRef: esv1beta1.ExternalSecretDataRemoteRef{
+								Key:      remoteKey,
+								Property: remoteProperty,
+							},
+						},
+					},
+				},
+			},
+		}
+	}
```

Comment:
With the testCase now having a `GenericStore` we should be able to get rid of all this duplication by reusing the code from the `Describe()` block above. We just need to modify the `tc.secretStore` becoming a `ClusterSecretStore`. Could you take a look at that if you have the time? :pray:

Reply:
Attempted to do what you described - tests still run well and the code looks better now, less duplicated.
Only thing I dislike now is the AfterEach block that has to check if the object exists before deleting - since we do not know which object was created in that scope.

If you have a better suggestion to that issue there I would love to hear.